docfx_yaml/extension.py

```diff
@@ -1417,58 +1417,127 @@ def find_markdown_pages(app, outdir):
+                not (visited_words and f"<a href=\"{hard_coded_references[keyword]}" in visited_words[-1]):
+                return f"<a href=\"{hard_coded_references[keyword]}\">{keyword}</a>"
+
+        else:
+            # If it is already processed as cross reference, skip over it.
+            if "<xref" not in words[index-1] and \
+                not (visited_words and f"<xref uid=\"{keyword}" in visited_words[-1]):
+                return f"<xref uid=\"{keyword}\">{keyword}</xref>"
+
+    return None
+
+
+def convert_cross_references(content: str, current_object_name: str, entry_names: List[str]):
+    """Finds and replaces references that should be a cross reference in given content.
+
+    This should not convert any references that contain `current_object_name`,
+    i.e. if we're processing docstring for `google.cloud.spanner.v1.services`,
+    references to `google.cloud.spanner.v1.services` should not be convereted
+    to references.
+
+    Args:
+        content: given body of content to parse and look for references
+        current_object_name: the name of the current Python object being looked at
+        entry_names: list of keyword references to look for
+
+    Returns:
+        content that has been modified with proper cross references if found.
+
+    """
     words = content.split(" ")
-    new_words = []
+
+    # Contains a list of words that is not a valid reference or converted
+    # references.
+    visited_words = []
+
+    # TODO(https://github.com/googleapis/sphinx-docfx-yaml/issues/208):
+    # remove this in the future.
+    iam_policy_link = "http://github.com/googleapis/python-grpc-google-iam-v1/blob/8e73b45993f030f521c0169b380d0fbafe66630b/google/iam/v1/iam_policy_pb2_grpc.py"
+    hard_coded_references = {
+        "google.iam.v1.iam_policy_pb2.SetIamPolicyRequest": iam_policy_link + "#L103-L109",
+        "google.iam.v1.iam_policy_pb2.GetIamPolicyRequest": iam_policy_link + "#L111-L118",
+        "google.iam.v1.iam_policy_pb2.TestIamPermissionsRequest": iam_policy_link + "#L120-L131",
+        "google.iam.v1.iam_policy_pb2.TestIamPermissionsResponse": iam_policy_link + "#L120-L131"
+    }
+    entry_names.extend(hard_coded_references.keys())
+
     # Using counter to check if the entry is already a cross reference.
     for index, word in enumerate(words):
-        cross_reference = ""
+        cross_reference = None
         for keyword in entry_names:
-            if keyword != current_name and keyword not in current_name and keyword in word:
-                # If it is already processed as cross reference, skip over it.
-                if "<xref" in words[index-1] or (new_words and f"<xref uid=\"{keyword}" in new_words[-1]):
-                    continue
-                cross_reference = f"<xref uid=\"{keyword}\">{keyword}</xref>"
-                new_words.append(word.replace(keyword, cross_reference))
+            # Do not convert references to itself.
+            if keyword in current_object_name:
```

Comment:
Apologies if this was asked before. Why is this `in` and not `==`? I'm imagining something like `foo.Foo` and `foo.FooBar`. I suppose there might be other punctuation or something?